### PR TITLE
Manifest validation

### PR DESCRIPTION
**Pull request by gaganhegde**
**What does does this PR do / why we need it**:

Entity names should be  Domain Name RFC 1035 compatible for Services, applications, and environments
Check for mandatory fields like webhook, etc
Every service should have a webhook ,Webhook secrets.
Check for the Domain Name RFC 1035 compatible name.
Check for duplicates:
Environment names must be unique
Application names must be unique in an environment (applications can be repeated across different environments)
Service names must be unique in an application (services can be repeated across different applications/environments)

> /kind design

**Which issue(s) this PR fixes**:
https://issues.redhat.com/browse/GITOPS-105

**How to test changes / Special notes to the reviewer**:
The test cases should be run using go test -v on the config repo in manifest.
The test yaml files are present in the testdata folder

### CODE REVIEW

File: pkg/manifest/build.go [hunk 1]
```diff
@@ -24,7 +25,10 @@ func BuildResources(o *BuildParameters) error {
 	if err != nil {
 		return fmt.Errorf("failed to parse manifest: %w", err)
 	}
-
+	if err := m.Validate(); err != nil {
+		log.Println("These are the errors", err)
```
Comment (bigkevmcd) [inline pkg/manifest/build.go]:
```suggestion
		log.Println("could not process the manifest, validation failed", err)
```

File: pkg/manifest/config/testdata/duplicate_application.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    pipelines:
+      integration:
+        template: dev-ci-template
+        binding: dev-ci-binding
+    apps:
+      - name: my-app-1 #Application my-app-1
+        services:
+          - name: app-1-service-http
+          - name: app-1-service-metrics
+      - name: my-app-1 #Dupliacte application my-app-1 (invalid)
```
Comment (bigkevmcd) [inline pkg/manifest/config/testdata/duplicate_application.yaml]:
```suggestion
      - name: my-app-1 # Duplicate application my-app-1 (invalid)
```

File: pkg/manifest/config/testdata/duplicate_application.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+environments:
+  - name: app-environment
+    pipelines:
+      integration:
+        template: dev-ci-template
+        binding: dev-ci-binding
+    apps:
+      - name: my-app-1 #Application my-app-1
```
Comment (bigkevmcd) [inline pkg/manifest/config/testdata/duplicate_application.yaml]:
```suggestion
      - name: my-app-1 # Application my-app-1
```

File: pkg/manifest/config/testdata/duplicate_environment.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+environments:
+  - name: duplicate-environment #Environment duplicate-environment
```
Comment (bigkevmcd) [inline pkg/manifest/config/testdata/duplicate_environment.yaml]:
```suggestion
  - name: duplicate-environment # Environment duplicate-environment 
```

File: pkg/manifest/config/testdata/duplicate_service.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+environments:
+  - name: duplicate-service
+    pipelines:
+      integration:
+        template: dev-ci-template
+        binding: dev-ci-binding
+    apps:
+      - name: my-app-1
+        services:
+          - name: app-1-service-http #Service Name
```
Comment (bigkevmcd) [inline pkg/manifest/config/testdata/duplicate_service.yaml]:
```suggestion
          - name: app-1-service-http # Service Name
```

File: pkg/manifest/config/testdata/duplicate_service.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+          - name: app-1-service-http #Repeated Service Name(invalid)
```
Comment (bigkevmcd) [inline pkg/manifest/config/testdata/duplicate_service.yaml]:
```suggestion
          - name: app-1-service-http # Duplicate Service Name (invalid)
```

File: pkg/manifest/build.go [hunk 7]
```diff
@@ -24,7 +25,10 @@ func BuildResources(o *BuildParameters) error {
 	if err != nil {
 		return fmt.Errorf("failed to parse manifest: %w", err)
 	}
-
+	if err := m.Validate(); err != nil {
+		log.Println("could not process the manifest, validation failed", err)
```
Comment (wtam2018) [inline pkg/manifest/build.go:29]: Why we write to the log in here only?

File: pkg/manifest/config/validate.go [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+	if app.Services == nil && app.ConfigRepo == nil {
+		vv.errs = append(vv.errs, missingFieldsError([]string{"services", "config_repo"}, []string{appPath}))
+	}
+	if app.Services != nil && app.ConfigRepo != nil {
+		vv.errs = append(vv.errs, apis.ErrMultipleOneOf(yamlJoin(appPath, "services"), yamlJoin(appPath, "config_repo")))
+	}
+	if app.ConfigRepo != nil {
+		vv.errs = append(vv.errs, validateConfigRepo(app.ConfigRepo, yamlJoin(appPath, "config_repo"))...)
+	}
```
Comment (wtam2018) [inline pkg/manifest/config/validate.go:58]: If application has configRepo, validation fails (regardless services exist or not)?

**Pull request merged**